Added DNS resolution and latency test for the Lobby servers.

pzhlkj6612 opened the pull request:
Thanks for your great work. I've made some modification about the process of connection to the Lobby servers. Please review and test it. Thank you.

### Commit message

1. Get all IP addresses of servers by resolving the hostname.
2. Test the latency to the servers and store them in a list.
3. If the latency is greater than 400 ms, ignore the server.
4. Sort the list and connect to the servers one by one.

### It can...

1. Find the closest and reachable server in few seconds.
2. Connect to all servers bound to one domain name.

### It also can...

1. Resolve #141 .

### Test

At 10:16 PM 4/17/2020, there are 843 players on the server.

<details>
  <summary>client.log</summary>
  <pre>
...
17.04. 22:16:59.701    Writing settings INI.
17.04. 22:16:59.739    Attempting to DNS resolve GameSurge (irc.gamesurge.net).
17.04. 22:16:59.760    Attempting to DNS resolve GameSurge London, UK (Burstfire.UK.EU.GameSurge.net).
17.04. 22:16:59.769    Attempting to DNS resolve GameSurge Chicago, IL (ColoCrossing.IL.US.GameSurge.net).
17.04. 22:16:59.780    DNS resolved GameSurge (irc.gamesurge.net): 108.174.48.100, 91.217.189.76, 192.223.27.109, 162.248.94.123, 195.68.206.250, 195.8.250.180
17.04. 22:16:59.789    DNS resolved GameSurge London, UK (Burstfire.UK.EU.GameSurge.net): 195.68.206.250
17.04. 22:16:59.804    Attempting to ping GameSurge (195.8.250.180).
17.04. 22:16:59.812    DNS resolved GameSurge Chicago, IL (ColoCrossing.IL.US.GameSurge.net): 108.174.48.100
17.04. 22:16:59.829    Attempting to ping GameSurge London, UK (195.68.206.250).
17.04. 22:16:59.840    Attempting to ping GameSurge Chicago, IL (108.174.48.100).
17.04. 22:17:00.064    The latency in milliseconds to the server GameSurge (195.8.250.180): 249.
17.04. 22:17:00.076    Attempting to ping GameSurge (195.68.206.250).
17.04. 22:17:00.140    The latency in milliseconds to the server GameSurge Chicago, IL (108.174.48.100): 286.
17.04. 22:17:00.151    Attempting to DNS resolve GameSurge Newark, NJ (Gameservers.NJ.US.GameSurge.net).
17.04. 22:17:00.176    DNS resolved GameSurge Newark, NJ (Gameservers.NJ.US.GameSurge.net): 208.167.237.120
17.04. 22:17:00.191    Attempting to ping GameSurge Newark, NJ (208.167.237.120).
17.04. 22:17:00.290    Pinging the server GameSurge London, UK (195.68.206.250) timed out!
17.04. 22:17:00.291    Pinging the server GameSurge (195.68.206.250) timed out!
17.04. 22:17:00.310    Pinging the server GameSurge Newark, NJ (208.167.237.120) timed out!
17.04. 22:17:00.319    Attempting to DNS resolve GameSurge Santa Ana, CA (Krypt.CA.US.GameSurge.net).
17.04. 22:17:00.330    Attempting to ping GameSurge (162.248.94.123).
17.04. 22:17:00.339    Attempting to DNS resolve GameSurge Seattle, WA (NuclearFallout.WA.US.GameSurge.net).
17.04. 22:17:00.351    DNS resolved GameSurge Santa Ana, CA (Krypt.CA.US.GameSurge.net): 209.11.244.82
17.04. 22:17:00.361    Attempting to ping GameSurge Santa Ana, CA (209.11.244.82).
17.04. 22:17:00.376    DNS resolved GameSurge Seattle, WA (NuclearFallout.WA.US.GameSurge.net): 162.248.94.123
17.04. 22:17:00.390    Attempting to ping GameSurge Seattle, WA (162.248.94.123).
17.04. 22:17:00.629    The latency in milliseconds to the server GameSurge Seattle, WA (162.248.94.123): 239.
17.04. 22:17:00.660    Attempting to DNS resolve GameSurge Stockholm, Sweden (Portlane.SE.EU.GameSurge.net).
17.04. 22:17:00.680    DNS resolved GameSurge Stockholm, Sweden (Portlane.SE.EU.GameSurge.net): 91.217.189.76
17.04. 22:17:00.681    Attempting to ping GameSurge Stockholm, Sweden (91.217.189.76).
17.04. 22:17:00.789    Pinging the server GameSurge (162.248.94.123) timed out!
17.04. 22:17:00.801    Pinging the server GameSurge Santa Ana, CA (209.11.244.82) timed out!
17.04. 22:17:00.821    Pinging the server GameSurge Stockholm, Sweden (91.217.189.76) timed out!
17.04. 22:17:00.821    Attempting to ping GameSurge (192.223.27.109).
17.04. 22:17:00.852    Attempting to DNS resolve GameSurge NYC, NY (Prothid.NY.US.GameSurge.Net).
17.04. 22:17:00.869    Attempting to DNS resolve GameSurge Wuppertal, Germany (TAL.DE.EU.GameSurge.net).
17.04. 22:17:00.877    DNS resolved GameSurge NYC, NY (Prothid.NY.US.GameSurge.Net): 192.223.27.109
17.04. 22:17:00.901    DNS resolved GameSurge Wuppertal, Germany (TAL.DE.EU.GameSurge.net): 195.8.250.180
17.04. 22:17:00.909    Attempting to ping GameSurge NYC, NY (192.223.27.109).
17.04. 22:17:00.921    Skipped a duplicate IP from GameSurge Wuppertal, Germany (195.8.250.180).
17.04. 22:17:00.934    Attempting to DNS resolve GameSurge IP 208.167.237.120 (208.167.237.120).
17.04. 22:17:00.950    DNS resolved GameSurge IP 208.167.237.120 (208.167.237.120): 208.167.237.120
17.04. 22:17:00.954    Attempting to ping GameSurge IP 208.167.237.120 (208.167.237.120).
17.04. 22:17:01.124    The latency in milliseconds to the server GameSurge (192.223.27.109): 272.
17.04. 22:17:01.140    Attempting to ping GameSurge (91.217.189.76).
17.04. 22:17:01.190    The latency in milliseconds to the server GameSurge NYC, NY (192.223.27.109): 268.
17.04. 22:17:01.201    Attempting to DNS resolve GameSurge IP 192.223.27.109 (192.223.27.109).
17.04. 22:17:01.214    DNS resolved GameSurge IP 192.223.27.109 (192.223.27.109): 192.223.27.109
17.04. 22:17:01.229    Skipped a duplicate IP from GameSurge IP 192.223.27.109 (192.223.27.109).
17.04. 22:17:01.244    Attempting to DNS resolve GameSurge IP 108.174.48.100 (108.174.48.100).
17.04. 22:17:01.260    Attempting to DNS resolve GameSurge IP 208.146.35.105 (208.146.35.105).
17.04. 22:17:01.260    DNS resolved GameSurge IP 108.174.48.100 (108.174.48.100): 108.174.48.100
17.04. 22:17:01.279    DNS resolved GameSurge IP 208.146.35.105 (208.146.35.105): 208.146.35.105
17.04. 22:17:01.291    Skipped a duplicate IP from GameSurge IP 108.174.48.100 (108.174.48.100).
17.04. 22:17:01.309    Pinging the server GameSurge IP 208.167.237.120 (208.167.237.120) timed out!
17.04. 22:17:01.314    Pinging the server GameSurge (91.217.189.76) timed out!
17.04. 22:17:01.329    Attempting to ping GameSurge IP 208.146.35.105 (208.146.35.105).
17.04. 22:17:01.334    Attempting to DNS resolve GameSurge IP 195.8.250.180 (195.8.250.180).
17.04. 22:17:01.350    Attempting to DNS resolve GameSurge IP 91.217.189.76 (91.217.189.76).
17.04. 22:17:01.361    Attempting to ping GameSurge (108.174.48.100).
17.04. 22:17:01.371    DNS resolved GameSurge IP 195.8.250.180 (195.8.250.180): 195.8.250.180
17.04. 22:17:01.380    DNS resolved GameSurge IP 91.217.189.76 (91.217.189.76): 91.217.189.76
17.04. 22:17:01.391    Skipped a duplicate IP from GameSurge IP 195.8.250.180 (195.8.250.180).
17.04. 22:17:01.411    Attempting to ping GameSurge IP 91.217.189.76 (91.217.189.76).
17.04. 22:17:01.420    Attempting to DNS resolve GameSurge IP 195.68.206.250 (195.68.206.250).
17.04. 22:17:01.432    DNS resolved GameSurge IP 195.68.206.250 (195.68.206.250): 195.68.206.250
17.04. 22:17:01.442    Attempting to ping GameSurge IP 195.68.206.250 (195.68.206.250).
17.04. 22:17:01.701    The latency in milliseconds to the server GameSurge IP 91.217.189.76 (91.217.189.76): 284.
17.04. 22:17:01.790    Pinging the server GameSurge IP 208.146.35.105 (208.146.35.105) timed out!
17.04. 22:17:01.800    Pinging the server GameSurge (108.174.48.100) timed out!
17.04. 22:17:01.820    Pinging the server GameSurge IP 195.68.206.250 (195.68.206.250) timed out!
17.04. 22:17:04.849    Attempting connection to 162.248.94.123:6667
17.04. 22:17:04.862    Connecting to 162.248.94.123 port 6667 timed out!
17.04. 22:17:07.121    Attempting connection to 162.248.94.123:5960
17.04. 22:17:07.130    Succesfully connected to 162.248.94.123 on port 5960
17.04. 22:17:07.140    Registering.
...
  </pre>
</details>

Knighthana commented:
That's really a big good change for the client. I really need this feature, on my restricted network.
Thanks a lot, sincerely.

pzhlkj6612 commented:
### Changes

- Re-based on the latest commit.
- Moved the code to `GetAvailableServerList` method.
- Added a log message about the number of available Lobby servers.

### Test

At 12:34 AM 7/25/2020, there are 1350 players on the server.

<details>
  <summary>client.log</summary>
  <pre>
...
25.07. 00:34:47.013    Writing settings INI.
25.07. 00:34:47.036    Attempting to DNS resolve GameSurge (irc.gamesurge.net).
25.07. 00:34:47.045    Attempting to DNS resolve GameSurge Seattle, WA (NuclearFallout.WA.US.GameSurge.net).
25.07. 00:34:47.053    Attempting to DNS resolve GameSurge Chicago, IL (ColoCrossing.IL.US.GameSurge.net).
25.07. 00:34:47.059    Attempting to DNS resolve GameSurge London, UK (Burstfire.UK.EU.GameSurge.net).
25.07. 00:34:47.065    Attempting to DNS resolve GameSurge Newark, NJ (Gameservers.NJ.US.GameSurge.net).
25.07. 00:34:47.072    Attempting to DNS resolve GameSurge Santa Ana, CA (Krypt.CA.US.GameSurge.net).
25.07. 00:34:47.078    Attempting to DNS resolve GameSurge Stockholm, Sweden (Portlane.SE.EU.GameSurge.net).
25.07. 00:34:47.102    DNS resolved GameSurge Seattle, WA (NuclearFallout.WA.US.GameSurge.net): 162.248.94.123
25.07. 00:34:47.109    DNS resolved GameSurge Chicago, IL (ColoCrossing.IL.US.GameSurge.net): 108.174.48.100
25.07. 00:34:47.116    Attempting to ping GameSurge Seattle, WA (162.248.94.123).
25.07. 00:34:47.121    Attempting to ping GameSurge Chicago, IL (108.174.48.100).
25.07. 00:34:47.130    DNS resolved GameSurge Santa Ana, CA (Krypt.CA.US.GameSurge.net): 209.11.244.82
25.07. 00:34:47.135    Attempting to ping GameSurge Santa Ana, CA (209.11.244.82).
25.07. 00:34:47.258    DNS resolved GameSurge (irc.gamesurge.net): 192.223.27.109, 108.174.48.100, 91.217.189.76, 162.248.94.123, 195.8.250.180, 195.68.206.250
25.07. 00:34:47.272    Attempting to ping GameSurge (195.68.206.250).
25.07. 00:34:47.290    DNS resolved GameSurge Newark, NJ (Gameservers.NJ.US.GameSurge.net): 208.167.237.120
25.07. 00:34:47.305    DNS resolved GameSurge London, UK (Burstfire.UK.EU.GameSurge.net): 195.68.206.250
25.07. 00:34:47.314    Attempting to ping GameSurge London, UK (195.68.206.250).
25.07. 00:34:47.321    Attempting to ping GameSurge Newark, NJ (208.167.237.120).
25.07. 00:34:47.424    The latency in milliseconds to the server GameSurge Chicago, IL (108.174.48.100): 295.
25.07. 00:34:47.432    Attempting to DNS resolve GameSurge NYC, NY (Prothid.NY.US.GameSurge.Net).
25.07. 00:34:47.580    Pinging the server GameSurge Seattle, WA (162.248.94.123) timed out!
25.07. 00:34:47.595    Attempting to DNS resolve GameSurge Wuppertal, Germany (TAL.DE.EU.GameSurge.net).
25.07. 00:34:47.605    Pinging the server GameSurge Santa Ana, CA (209.11.244.82) timed out!
25.07. 00:34:47.611    Pinging the server GameSurge London, UK (195.68.206.250) timed out!
25.07. 00:34:47.616    Pinging the server GameSurge Newark, NJ (208.167.237.120) timed out!
25.07. 00:34:47.623    Pinging the server GameSurge (195.68.206.250) timed out!
25.07. 00:34:47.628    Attempting to DNS resolve GameSurge IP 208.167.237.120 (208.167.237.120).
25.07. 00:34:47.633    Attempting to DNS resolve GameSurge IP 192.223.27.109 (192.223.27.109).
25.07. 00:34:47.639    Attempting to DNS resolve GameSurge IP 108.174.48.100 (108.174.48.100).
25.07. 00:34:47.645    DNS resolved GameSurge Wuppertal, Germany (TAL.DE.EU.GameSurge.net): 195.8.250.180
25.07. 00:34:47.650    Attempting to ping GameSurge (195.8.250.180).
25.07. 00:34:47.658    DNS resolved GameSurge IP 208.167.237.120 (208.167.237.120): 208.167.237.120
25.07. 00:34:47.663    DNS resolved GameSurge IP 192.223.27.109 (192.223.27.109): 192.223.27.109
25.07. 00:34:47.669    DNS resolved GameSurge IP 108.174.48.100 (108.174.48.100): 108.174.48.100
25.07. 00:34:47.676    Attempting to ping GameSurge Wuppertal, Germany (195.8.250.180).
25.07. 00:34:47.681    DNS resolved GameSurge NYC, NY (Prothid.NY.US.GameSurge.Net): 192.223.27.109
25.07. 00:34:47.687    Attempting to ping GameSurge IP 208.167.237.120 (208.167.237.120).
25.07. 00:34:47.693    Attempting to ping GameSurge IP 192.223.27.109 (192.223.27.109).
25.07. 00:34:47.698    Skipped a duplicate IP from GameSurge IP 108.174.48.100 (108.174.48.100).
25.07. 00:34:47.704    Attempting to ping GameSurge NYC, NY (192.223.27.109).
25.07. 00:34:47.711    Attempting to DNS resolve GameSurge IP 208.146.35.105 (208.146.35.105).
25.07. 00:34:47.716    DNS resolved GameSurge IP 208.146.35.105 (208.146.35.105): 208.146.35.105
25.07. 00:34:47.722    Attempting to ping GameSurge IP 208.146.35.105 (208.146.35.105).
25.07. 00:34:47.899    DNS resolved GameSurge Stockholm, Sweden (Portlane.SE.EU.GameSurge.net): 91.217.189.76
25.07. 00:34:47.905    Attempting to ping GameSurge Stockholm, Sweden (91.217.189.76).
25.07. 00:34:47.930    The latency in milliseconds to the server GameSurge Wuppertal, Germany (195.8.250.180): 248.
25.07. 00:34:47.936    Attempting to DNS resolve GameSurge IP 195.8.250.180 (195.8.250.180).
25.07. 00:34:47.943    DNS resolved GameSurge IP 195.8.250.180 (195.8.250.180): 195.8.250.180
25.07. 00:34:47.951    Skipped a duplicate IP from GameSurge IP 195.8.250.180 (195.8.250.180).
25.07. 00:34:47.958    Attempting to DNS resolve GameSurge IP 91.217.189.76 (91.217.189.76).
25.07. 00:34:47.966    DNS resolved GameSurge IP 91.217.189.76 (91.217.189.76): 91.217.189.76
25.07. 00:34:47.974    Attempting to ping GameSurge IP 91.217.189.76 (91.217.189.76).
25.07. 00:34:47.987    The latency in milliseconds to the server GameSurge NYC, NY (192.223.27.109): 274.
25.07. 00:34:47.998    Attempting to DNS resolve GameSurge IP 195.68.206.250 (195.68.206.250).
25.07. 00:34:48.007    DNS resolved GameSurge IP 195.68.206.250 (195.68.206.250): 195.68.206.250
25.07. 00:34:48.016    Attempting to ping GameSurge IP 195.68.206.250 (195.68.206.250).
25.07. 00:34:48.080    Pinging the server GameSurge (195.8.250.180) timed out!
25.07. 00:34:48.088    Attempting to ping GameSurge (162.248.94.123).
25.07. 00:34:48.095    Pinging the server GameSurge IP 195.68.206.250 (195.68.206.250) timed out!
25.07. 00:34:48.100    Pinging the server GameSurge IP 208.146.35.105 (208.146.35.105) timed out!
25.07. 00:34:48.105    Pinging the server GameSurge IP 192.223.27.109 (192.223.27.109) timed out!
25.07. 00:34:48.110    Pinging the server GameSurge Stockholm, Sweden (91.217.189.76) timed out!
25.07. 00:34:48.117    Pinging the server GameSurge IP 208.167.237.120 (208.167.237.120) timed out!
25.07. 00:34:48.123    Pinging the server GameSurge IP 91.217.189.76 (91.217.189.76) timed out!
25.07. 00:34:48.130    Attempting to ping GameSurge (192.223.27.109).
25.07. 00:34:48.138    Attempting to ping GameSurge (108.174.48.100).
25.07. 00:34:48.144    Attempting to ping GameSurge (91.217.189.76).
25.07. 00:34:48.506    The latency in milliseconds to the server GameSurge (91.217.189.76): 344.
25.07. 00:34:48.579    Pinging the server GameSurge (162.248.94.123) timed out!
25.07. 00:34:48.593    Pinging the server GameSurge (108.174.48.100) timed out!
25.07. 00:34:48.603    Pinging the server GameSurge (192.223.27.109) timed out!
25.07. 00:34:48.615    The number of available Lobby servers is 4.
25.07. 00:34:48.883    Attempting connection to 195.8.250.180:6660
25.07. 00:34:48.890    Succesfully connected to 195.8.250.180 on port 6660
25.07. 00:34:48.900    Registering.
...
  </pre>
</details>

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+                    foreach (var innerEx in ((AggregateException)ex).Flatten().InnerExceptions)
+                    {
+                        Logger.Log("Unable to contact with the server. " + innerEx.Message);
+                    }
```
Would be better to have no braces here to comply with client code style.

pzhlkj6612 replied:
Fixed in c448a0473f5abb25e469f6ee78691bace0932b62 .

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+                else
+                {
+                    Logger.Log("Unable to contact with the server. " + ex.Message);
+                }
```
Ditto.

pzhlkj6612 replied:
Fixed in c448a0473f5abb25e469f6ee78691bace0932b62 .

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+                            if (availableServerAndLatencyDict.Any(item => item.Key.Host == theIPAddress.ToString()))
+                            {
+                                Logger.Log($"Skipped a duplicate IP from {serverName} ({theIPAddress}).");
+
```
Additional newlines are unnecessary here, better to keep the code vertically compact.

pzhlkj6612 replied:
Fixed in c448a0473f5abb25e469f6ee78691bace0932b62 .

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+                            if (failedServerIPs.Contains(theIPAddress.ToString()))
+                            {
+                                Logger.Log($"Skipped a failed server {serverName} ({theIPAddress}).");
+
```
Ditto. Not mentioning unnecessary newlines from now on, there are a too much of them.

pzhlkj6612 replied:
Fixed in c448a0473f5abb25e469f6ee78691bace0932b62 .

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+
+                        foreach (var serverIPAddress in serverIPAddresses)
+                        {
+                            var theIPAddress = serverIPAddress;
```
Looks like an absolutely unnecessary local variable.

pzhlkj6612 replied:
Fixed in 4d3608cc23dc908ee2b838dcfbf9b2b335a52e97 .

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+                                    {
+                                        Logger.Log($"Pinging the server {serverName} ({theIPAddress}) timed out!");
+                                    }
+                                }
```
Those 4 braces can be omitted. You should use `else if` here and write the logger call in a single line without braces.

pzhlkj6612 replied:
Fixed in c448a0473f5abb25e469f6ee78691bace0932b62 .

Metadorius commented on DXMainClient/Online/Connection.cs:
```diff
@@ -324,6 +326,110 @@ private void HandleComm(object client)
+                {
+                    var serverHostnameOrIPAddress = server.Host;
+                    var serverName = server.Name;
+                    var serverPorts = server.Ports;
```
Try to avoid using `var` on every ocassion, it's clashing with client code style.

pzhlkj6612 replied:
Fixed in c09827cbb9556f93fc5acf67448ece2194546faa .

Metadorius reviewed (changes requested):
The changes look good aside from styling issues. Sorry for the delay in reviewing the PR.

pzhlkj6612 commented:
Hello @Metadorius , thank you very much for your review.

---

After applying these changes, I tested it again.

At 3:17 PM 9/19/2020, there are 623 players on the server.

<details>
  <summary>client.log</summary>
  <pre>
...
19.09. 15:17:11.085    Writing settings INI.
19.09. 15:17:11.132    Attempting to DNS resolve GameSurge (irc.gamesurge.net).
19.09. 15:17:11.155    Attempting to DNS resolve GameSurge London, UK (Burstfire.UK.EU.GameSurge.net).
19.09. 15:17:11.204    Attempting to DNS resolve GameSurge Chicago, IL (ColoCrossing.IL.US.GameSurge.net).
19.09. 15:17:11.217    DNS resolved GameSurge (irc.gamesurge.net): 108.174.48.100, 192.223.27.109, 195.68.206.250, 195.8.250.180, 91.217.189.76, 162.248.94.123
19.09. 15:17:11.242    DNS resolved GameSurge London, UK (Burstfire.UK.EU.GameSurge.net): 195.68.206.250
19.09. 15:17:11.262    DNS resolved GameSurge Chicago, IL (ColoCrossing.IL.US.GameSurge.net): 108.174.48.100
19.09. 15:17:11.267    Attempting to ping GameSurge (162.248.94.123).
19.09. 15:17:11.273    Skipped a duplicate IP from GameSurge London, UK (195.68.206.250).
19.09. 15:17:11.283    Skipped a duplicate IP from GameSurge Chicago, IL (108.174.48.100).
19.09. 15:17:11.312    Attempting to DNS resolve GameSurge Newark, NJ (Gameservers.NJ.US.GameSurge.net).
19.09. 15:17:11.317    Attempting to DNS resolve GameSurge Santa Ana, CA (Krypt.CA.US.GameSurge.net).
19.09. 15:17:11.337    DNS resolved GameSurge Newark, NJ (Gameservers.NJ.US.GameSurge.net): 208.167.237.120
19.09. 15:17:11.362    DNS resolved GameSurge Santa Ana, CA (Krypt.CA.US.GameSurge.net): 209.11.244.82
19.09. 15:17:11.364    Attempting to ping GameSurge Newark, NJ (208.167.237.120).
19.09. 15:17:11.383    Attempting to ping GameSurge Santa Ana, CA (209.11.244.82).
19.09. 15:17:11.503    The latency in milliseconds to the server GameSurge (162.248.94.123): 223.
19.09. 15:17:11.533    Attempting to ping GameSurge (91.217.189.76).
19.09. 15:17:11.542    Failed to ping the server GameSurge Newark, NJ (208.167.237.120): TimedOut.
19.09. 15:17:11.553    Failed to ping the server GameSurge Santa Ana, CA (209.11.244.82): TimedOut.
19.09. 15:17:11.564    Failed to ping the server GameSurge (91.217.189.76): TimedOut.
19.09. 15:17:11.584    Attempting to DNS resolve GameSurge Seattle, WA (NuclearFallout.WA.US.GameSurge.net).
19.09. 15:17:11.602    Attempting to DNS resolve GameSurge Stockholm, Sweden (Portlane.SE.EU.GameSurge.net).
19.09. 15:17:11.604    Attempting to ping GameSurge (195.8.250.180).
19.09. 15:17:11.624    DNS resolved GameSurge Seattle, WA (NuclearFallout.WA.US.GameSurge.net): 162.248.94.123
19.09. 15:17:11.633    DNS resolved GameSurge Stockholm, Sweden (Portlane.SE.EU.GameSurge.net): 91.217.189.76
19.09. 15:17:11.655    Skipped a duplicate IP from GameSurge Seattle, WA (162.248.94.123).
19.09. 15:17:11.667    Skipped a duplicate IP from GameSurge Stockholm, Sweden (91.217.189.76).
19.09. 15:17:11.683    Attempting to DNS resolve GameSurge NYC, NY (Prothid.NY.US.GameSurge.Net).
19.09. 15:17:11.704    Attempting to DNS resolve GameSurge Wuppertal, Germany (TAL.DE.EU.GameSurge.net).
19.09. 15:17:11.723    DNS resolved GameSurge NYC, NY (Prothid.NY.US.GameSurge.Net): 192.223.27.109
19.09. 15:17:11.732    Skipped a duplicate IP from GameSurge NYC, NY (192.223.27.109).
19.09. 15:17:11.744    DNS resolved GameSurge Wuppertal, Germany (TAL.DE.EU.GameSurge.net): 195.8.250.180
19.09. 15:17:11.765    Attempting to DNS resolve GameSurge IP 208.167.237.120 (208.167.237.120).
19.09. 15:17:11.777    Skipped a duplicate IP from GameSurge Wuppertal, Germany (195.8.250.180).
19.09. 15:17:11.786    DNS resolved GameSurge IP 208.167.237.120 (208.167.237.120): 208.167.237.120
19.09. 15:17:11.798    Attempting to DNS resolve GameSurge IP 192.223.27.109 (192.223.27.109).
19.09. 15:17:11.807    Skipped a duplicate IP from GameSurge IP 208.167.237.120 (208.167.237.120).
19.09. 15:17:11.823    DNS resolved GameSurge IP 192.223.27.109 (192.223.27.109): 192.223.27.109
19.09. 15:17:11.835    Attempting to DNS resolve GameSurge IP 108.174.48.100 (108.174.48.100).
19.09. 15:17:11.843    Skipped a duplicate IP from GameSurge IP 192.223.27.109 (192.223.27.109).
19.09. 15:17:11.852    DNS resolved GameSurge IP 108.174.48.100 (108.174.48.100): 108.174.48.100
19.09. 15:17:11.872    Attempting to DNS resolve GameSurge IP 208.146.35.105 (208.146.35.105).
19.09. 15:17:11.882    Skipped a duplicate IP from GameSurge IP 108.174.48.100 (108.174.48.100).
19.09. 15:17:11.890    The latency in milliseconds to the server GameSurge (195.8.250.180): 250.
19.09. 15:17:11.906    DNS resolved GameSurge IP 208.146.35.105 (208.146.35.105): 208.146.35.105
19.09. 15:17:11.913    Attempting to DNS resolve GameSurge IP 195.8.250.180 (195.8.250.180).
19.09. 15:17:11.922    Attempting to ping GameSurge (195.68.206.250).
19.09. 15:17:11.942    Attempting to ping GameSurge IP 208.146.35.105 (208.146.35.105).
19.09. 15:17:11.947    DNS resolved GameSurge IP 195.8.250.180 (195.8.250.180): 195.8.250.180
19.09. 15:17:11.954    Skipped a duplicate IP from GameSurge IP 195.8.250.180 (195.8.250.180).
19.09. 15:17:11.954    Attempting to DNS resolve GameSurge IP 91.217.189.76 (91.217.189.76).
19.09. 15:17:11.974    DNS resolved GameSurge IP 91.217.189.76 (91.217.189.76): 91.217.189.76
19.09. 15:17:11.992    Skipped a duplicate IP from GameSurge IP 91.217.189.76 (91.217.189.76).
19.09. 15:17:11.992    Attempting to DNS resolve GameSurge IP 195.68.206.250 (195.68.206.250).
19.09. 15:17:12.012    DNS resolved GameSurge IP 195.68.206.250 (195.68.206.250): 195.68.206.250
19.09. 15:17:12.017    Skipped a duplicate IP from GameSurge IP 195.68.206.250 (195.68.206.250).
19.09. 15:17:12.024    Attempting to ping GameSurge (108.174.48.100).
19.09. 15:17:12.042    Failed to ping the server GameSurge (195.68.206.250): TimedOut.
19.09. 15:17:12.054    Failed to ping the server GameSurge IP 208.146.35.105 (208.146.35.105): TimedOut.
19.09. 15:17:12.062    Failed to ping the server GameSurge (108.174.48.100): TimedOut.
19.09. 15:17:12.073    Attempting to ping GameSurge (192.223.27.109).
19.09. 15:17:12.373    The latency in milliseconds to the server GameSurge (192.223.27.109): 293.
19.09. 15:17:12.383    The number of available Lobby servers is 3.
19.09. 15:17:12.643    Attempting connection to 162.248.94.123:6667
19.09. 15:17:12.643    Succesfully connected to 162.248.94.123 on port 6667
19.09. 15:17:12.652    Registering.
...
  </pre>
</details>

Metadorius reviewed (approved):
I tested the changes on my machine, they seem to work as intended too. Also did a bit more adjustments to the code styling but since the PR doesn't allow pushes from maintainers I'll merge the PR first and then push to master.

pzhlkj6612 commented:
> since the PR doesn't allow pushes from maintainers...

Oh, maybe I accidentally closed this permission. 

>  Also did a bit more adjustments to the code styling...

Thank you, @Metadorius . I learned a lot from your commit (810ed48dc5123919d55de41a9c2925bbe77da3ea) and previous review:

- Lines should not be too long.
- Except for the innermost single-line statement, curly braces are necessary. At the same time, symmetry is also important.
- To provide a pause for the reader, add newlines before and after any multi-line structures.
- To make the _return_ statement more obvious, add a newline before it.

Metadorius commented:
@pzhlkj6612 I appreciate the detailed response, looking forward to your further contributions if you're willing to do any ;)
